examples/communities/louvain.py

```diff
@@ -1,33 +1,56 @@
-"""This module provides the louvain community detection example adapted from Cade et al.'s 2022 community detection paper"""
+"""This module provides a classical louvain community detection example adapted from Cade et al.'s 2022 community detection paper"""
 
 from networkx.algorithms.community import quality
 import networkx as nx
+from abc import ABC
 from typing import Optional
 
 
-class Louvain:
+class Louvain(ABC):
     """
     This class implements the classical Louvain algorithm like Cade et al. do in their community detection paper.
     It is initialized using an undirected networkx graph instance without selfloops.
     """
 
-    def __init__(self, G, *, keep_history=False) -> None:
+    def __init__(self, G, *, keep_history: bool = False) -> None:
         """Initialize Louvain algorithm based on a graph instance
 
         Args:
             G (nx.Graph): The initial graph where communities should be detected.
-            keep_history (bool): Keep maintaining a list of adjacency matrices and community mappings. Defaults to False.
+            keep_history: Keep maintaining a list of adjacency matrices and community mappings. Defaults to False.
         """
         self.keep_history = keep_history
         self.history = []
         self.update_graph(G)
-        self.C = self.single_partition()
+        self.C = self.single_partition()  # node to community map
+
+        # caches for `Sigma(u)`, `S(u, v)` and node strengths `strength(u)`
+        self.cache_Sigma: dict[int, int] = {}
+        self.cache_S: dict[(int, int), int] = {}
+        self.cache_strength: dict[int, int] = {}
 
     def update_graph(self, new_graph):
         self.G = nx.relabel.convert_node_labels_to_integers(new_graph)
```

Comment:
type missing. should be defined in constructor.

Reply:
Seems like you resolved this in your commit. I'll resolve these comments for now

Reply:
I was referring to `self.G` (and `self.A, self.W`) here. They should be declared in the constructor/class, with the appropriate type.